3.0/m_opmoderated.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2010 Daniel De Graaf <danieldg@inspircd.org>
+ *
+ * This file is part of InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/// $ModAuthor: genius3000
+/// $ModAuthorMail: genius3000@g3k.solutions
+/// $ModDepends: core 3.0
+/// $ModDesc: Implements channel mode +U and extban 'u' - moderator mute
+
+
+#include "inspircd.h"
+#include "modules/exemption.h"
+
+class ModuleOpModerated : public Module
+{
+	SimpleChannelModeHandler opmod;
+	CheckExemption::EventProvider exemptionprov;
+
+ public:
+	ModuleOpModerated()
+		: opmod(this, "opmoderated", 'U')
+		, exemptionprov(this)
+	{
+	}
+
+	void Prioritize() CXX11_OVERRIDE
+	{
+		// since we steal the message, we should be last (let everyone else eat it first)
+		ServerInstance->Modules->SetPriority(this, I_OnUserPreMessage, PRIORITY_LAST);
+	}
+
+	ModResult OnUserPreMessage(User *user, const MessageTarget& target, MessageDetails& details) CXX11_OVERRIDE
+	{
+		if (target.type != MessageTarget::TYPE_CHANNEL)
+			return MOD_RES_PASSTHRU;
+
+		if (!IS_LOCAL(user) || target.status)
+			return MOD_RES_PASSTHRU;
+
+		if (user->HasPrivPermission("channels/ignore-opmoderated"))
+			return MOD_RES_PASSTHRU;
+
+		Channel* const chan = target.Get<Channel>();
+		if (CheckExemption::Call(exemptionprov, user, chan, "opmoderated") == MOD_RES_ALLOW)
+			return MOD_RES_PASSTHRU;
+
+		if (!chan->GetExtBanStatus(user, 'u').check(!chan->IsModeSet(&opmod)) && chan->GetPrefixValue(user) < VOICE_VALUE)
+		{
+			MessageTarget newtarget = target;
+			newtarget.status = '@';
```

Comment:
Not sure if there's a better way to do this, without skipping the two On* calls below and sending to remote servers like timedbans does.

Reply:
You should be able to update the status in the existing message target.

Alternately, there's a CUList in MessageDetails that you could add all unprivileged users to.